packages/strapi-database/lib/validation/check-reserved-names.js

```diff
@@ -0,0 +1,77 @@
+'use strict';
+
+const _ = require('lodash');
+const constants = require('../constants');
+
+class ModelError extends Error {
+  constructor(message) {
+    super(message);
+    this.stack = null;
+  }
+}
+
+const checkReservedAttributeNames = (model, { manager }) => {
+  const reservedNames = [...constants.RESERVED_ATTRIBUTE_NAMES];
+
+  if (_.has(model, 'options.timestamps')) {
+    const [connectorCreatedAt, connectorUpdatedAt] = manager.connectors.getByConnection(
+      model.connection
+    ).defaultTimestamps;
+
+    if (Array.isArray(model.options.timestamps)) {
+      const [
+        createdAtAttribute = connectorCreatedAt,
+        updatedAtAttribute = connectorUpdatedAt,
+      ] = model.options.timestamps;
+
+      reservedNames.push(createdAtAttribute, updatedAtAttribute);
+    } else {
+      reservedNames.push(connectorCreatedAt, connectorUpdatedAt);
+    }
+  }
+
+  const usedReservedAttributeNames = _.intersection(Object.keys(model.attributes), reservedNames);
+
+  if (usedReservedAttributeNames.length > 0) {
+    throw new ModelError(
+      `Model "${
+        model.modelName
+      }" is using reserved attribute names "${usedReservedAttributeNames.join(
+        ', '
+      )}".\n-> Make sure you are not using a reserved name or overriding the defined timestamp attributes.`
+    );
+  }
+};
+
+const checkReservedModelName = model => {
+  if (constants.RESERVED_MODEL_NAMES.includes(model.modelName)) {
+    throw new ModelError(
+      `"${model.modelName}" is a reserved model name. You need to rename your model and the files associated with it.`
+    );
+  }
+};
+
+/**
+ * Checks that there are no model using reserved names (content type, component, attributes)
+ */
+module.exports = ({ strapi, manager }) => {
+  Object.keys(strapi.api).forEach(apiName => {
+    const api = strapi.api[apiName];
+
+    const models = api.models ? Object.values(api.models) : [];
+    models.forEach(model => {
+      checkReservedModelName(model);
+      checkReservedAttributeNames(model, { manager });
+    });
+  });
+
+  Object.keys(strapi.plugins).forEach(pluginName => {
+    const plugin = strapi.plugins[pluginName];
+
+    const models = plugin.models ? Object.values(plugin.models) : [];
+    models.forEach(model => {
```

Comment:
Duplicate block of L.62-65, maybe consider grouping all the models first and then apply the checks?

Reply:
I'm not sure this would be more readable 

```js
const apiModels = Object.values(strapi.api)
  .map(api => (api.models ? Object.values(api.models) : []))
  .reduce((acc, models) => acc.concat(models), []);

const pluginModels = Object.values(strapi.plugin)
  .map(plugin => (plugin.models ? Object.values(plugin.models) : []))
  .reduce((acc, models) => acc.concat(models), []);

apiModels.concat(pluginModels).forEach(model => {
  checkReservedModelName(model);
  checkReservedAttributeNames(model, { manager });
});
```

Reply:
Yeah, you're right...
What about doing something like
``` javascript
const applyChecks = models => {
    models.forEach(model => {
        checkReservedModelName(model);
        checkReservedAttributeNames(model, { manager });
    });
}
```

And then call it each time you need it?

Reply:
I made a change let me know if you like it ;) I would really like to use lodash/fp or ramda inside of the non FP lodash in the future

Reply:
Nice one!
Why not harmonizing it a bit more?
```javascript
const getModelsFrom = source => _.flatMap(source, iteratee => _.values(iteratee.model));

const applyChecks = model => {
    checkReservedModelName(model);
    checkReservedAttributeNames(model, { manager });
};

_.forEach(_.flatMap([strapi.api, strapi.plugins], getModelsFrom), applyChecks);
// or _.flatMap([strapi.api, strapi.plugins], getModelsFrom).forEach(applyChecks);
```